apis/Google.Cloud.Metadata.V1/Google.Cloud.Metadata.V1/MetadataClientImpl.cs

```diff
@@ -0,0 +1,502 @@
+        {
+            lock (eventLock)
+            {
+                if (watcher == null)
+                {
+                    watcher = new ValueWatcher(this, path, changeAction);
+                }
+
+                watcher.Event = Delegate.Combine(watcher.Event, value);
+            }
+        }
+
+        private void UnhookChangeEvent(ref ValueWatcher watcher, Delegate value)
+        {
+            lock (eventLock)
+            {
+                if (watcher == null)
+                {
+                    return;
+                }
+
+                watcher.Event = Delegate.Remove(watcher.Event, value);
+                if (watcher.Event == null)
+                {
+                    watcher.Stop();
+                    watcher = null;
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        public override event EventHandler InstanceMetadataChanged
+        {
+            add
+            {
+                HookChangeEvent(ref instanceWatcher, value, "instance?recursive=true", (sender, newValue) =>
+                {
+                    var temp = sender.Event as EventHandler;
+                    temp?.Invoke(this, EventArgs.Empty);
+                });
+            }
+            remove
+            {
+                UnhookChangeEvent(ref instanceWatcher, value);
+            }
+        }
+
+        /// <inheritdoc/>
+        public override event EventHandler<MaintenanceStatus> MaintenanceStatusChanged
+        {
+            add
+            {
+                HookChangeEvent(ref maintenanceStatusWatcher, value, "instance/maintenance-event", (sender, newValue) =>
+                {
+                    var temp = sender.Event as EventHandler<MaintenanceStatus>;
+                    temp?.Invoke(this, GetMaintenanceStatus(newValue));
+                });
+            }
+            remove
+            {
+                UnhookChangeEvent(ref maintenanceStatusWatcher, value);
+            }
+        }
+
+        /// <inheritdoc/>
+        public override event EventHandler ProjectMetadataChanged
+        {
+            add
+            {
+                HookChangeEvent(ref projectWatcher, value, "project?recursive=true", (sender, newValue) =>
+                {
+                    var temp = sender.Event as EventHandler;
+                    temp?.Invoke(this, EventArgs.Empty);
+                });
+            }
+            remove
+            {
+                UnhookChangeEvent(ref projectWatcher, value);
+            }
+        }
+
+        /// <inheritdoc/>
+        public override string WaitForChange(string key, TimeSpan timeout = default(TimeSpan)) => GetResult(token => WaitForChangeAsync(key, timeout, token));
+
+        /// <inheritdoc/>
+        public override async Task<string> WaitForChangeAsync(string key, TimeSpan timeout = default(TimeSpan), CancellationToken cancellationToken = default(CancellationToken))
```

Comment:
Ok good point. So for now I think the signature is fine. I'll just add validation logic to make sure parameters aren't specified. If/when I add ETag support, I'll add another parameter for that.